tests/unit/test_internal_utils.py

```diff
@@ -803,3 +804,75 @@ def test_get_dataset_type__conflicting_data():
         else:
             with pytest.raises(PowerGridError):
                 get_dataset_type(data=data)
+
+
+@pytest.fixture
+def row_dense_data():
+    source = initialize_array(DatasetType.update, ComponentType.source, (2, 3))
+    source["id"] = [[0, 2, 3], [0, 2, 3]]
+    source["u_ref"] = [[0.1, 0.2, 0.3], [0.4, 0.5, 0.6]]
+
+    sym_load = initialize_array(DatasetType.update, ComponentType.sym_load, (2, 1))
+    sym_load["id"] = [[1], [1]]
+    sym_load["p_specified"] = [[100.0], [200.0]]
+
+    return {
+        ComponentType.source: source,
+        ComponentType.sym_load: sym_load,
+    }
+
+
+@pytest.fixture
+def row_sparse_data():
+    transformer = initialize_array(DatasetType.update, ComponentType.transformer, 1)
+    transformer["id"] = 1
+    transformer["tap_pos"] = 3
+
+    sym_gen = initialize_array(DatasetType.update, ComponentType.sym_gen, (2, 4))
+    sym_gen["id"] = [[5, 6, 7, 8], [5, 6, 7, 8]]
+    sym_gen["q_specified"] = [[1.1, 2.2, 3.3, 4.4], [4.4, 3.3, 2.2, 1.1]]
+
+    return {
+        ComponentType.transformer: {
+            "data": transformer,
+            "indptr": np.array([0, 1, 1]),
+        },
+        ComponentType.sym_gen: {
+            "data": sym_gen,
+            "indptr": np.array([0, 1, 2]),
+        },
+    }
+
+
+@pytest.fixture(params=["row_dense_data", "row_sparse_data"])
+def row_data(request):
+    return request.getfixturevalue(request.param)
+
+
+def compare_row_data(actual_row_data, desired_row_data):
+    assert actual_row_data.keys() == desired_row_data.keys()
+    for comp_name in actual_row_data.keys():
+        actual_component = actual_row_data[comp_name]
+        desired_component = desired_row_data[comp_name]
+        if is_sparse(actual_component):
+            assert actual_component.keys() == desired_component.keys()
+            assert np.array_equal(actual_component["indptr"], desired_component["indptr"])
+            actual_component = actual_component["data"]
+            desired_component = desired_component["data"]
+        assert actual_component.dtype == desired_component.dtype
+        assert actual_component.shape == desired_component.shape
+
+        for field in actual_component.dtype.names:
+            assert np.all(
+                (np.isnan(actual_component[field]) & np.isnan(desired_component[field]))
+                | (actual_component[field] == desired_component[field])
+            )
```

Comment:
This condition seems sketchy. it shall pass when all nans match.

Try in the way of https://github.com/PowerGridModel/power-grid-model/blob/main/tests/unit/test_serialization.py#L779